sqlx-core/src/pool/inner.rs

```diff
@@ -344,31 +303,68 @@ where
     spawn(async move {
         while !pool.is_closed.load(Ordering::Acquire) {
             // reap at most the current size minus the minimum idle
-            let max_reaped = pool
-                .size
-                .load(Ordering::Acquire)
-                .saturating_sub(pool.options.min_size);
+            let max_reaped = pool.size().saturating_sub(pool.options.min_size);
 
             // collect connections to reap
             let (reap, keep) = (0..max_reaped)
                 // only connections waiting in the queue
-                .filter_map(|_| pool.idle.pop().ok())
+                .filter_map(|_| pool.pop_idle())
                 .partition::<Vec<_>, _>(|conn| {
-                    is_beyond_idle(conn, &pool.options)
-                        || is_beyond_lifetime(&conn.live, &pool.options)
+                    is_beyond_idle(conn, &pool.options) || is_beyond_lifetime(conn, &pool.options)
                 });
 
             for conn in keep {
                 // return these connections to the pool first
-                pool.idle.push(conn).expect("unreachable: pool overflowed");
+                pool.idle_conns
+                    .push(conn.into_leakable())
+                    .expect("BUG: connection queue overflow in spawn_reaper");
             }
 
             for conn in reap {
-                conn.close().await;
-                pool.size.fetch_sub(1, Ordering::AcqRel);
+                let _ = conn.close().await;
             }
 
             sleep(period).await;
         }
     });
 }
+
+/// RAII guard returned by `Pool::try_increment_size()` and others.
+///
+/// Will decrement the pool size if dropped, to avoid semantically "leaking" connections
+/// (where the pool thinks it has more connections than it does).
+pub(in crate::pool) struct DecrementSizeGuard<'a> {
+    size: &'a AtomicU32,
+    waiters: &'a SegQueue<Waker>,
+    dropped: bool,
```

Comment:
I love this solution :) But is a dropped flag really necessary? I didn't think it was possible to drop something twice in Rust (I don't think you can call the `.drop()` method manually, and the prelude `drop` function borrows its input, so attempts to reference it further wouldn't compile).